stdlib/public/core/CollectionDifference.swift

```diff
@@ -0,0 +1,378 @@
+  public let insertions: [Change]
+  
+  /// The `.remove` changes contained by this difference, from lowest offset to highest
+  public let removals: [Change]
+
+  /// The public initializer calls this function to ensure that its parameter
+  /// meets the conditions set in its documentation.
+  ///
+  /// - Parameter changes: a collection of `CollectionDifference.Change`
+  ///   instances intended to represent a valid state transition for
+  ///   `CollectionDifference`.
+  ///
+  /// - Returns: whether the parameter meets the following criteria:
+  ///
+  ///   1. All insertion offsets are unique
+  ///   2. All removal offsets are unique
+  ///   3. All associations between insertions and removals are symmetric
+  ///
+  /// Complexity: O(`changes.count`)
+  private static func _validateChanges<Changes: Collection>(
+    _ changes : Changes
+  ) -> Bool where Changes.Element == Change {
+    if changes.count == 0 { return true }
+
+    var insertAssocToOffset = Dictionary<Int,Int>()
+    var removeOffsetToAssoc = Dictionary<Int,Int>()
+    var insertOffset = Set<Int>()
+    var removeOffset = Set<Int>()
+
+    for c in changes {
+      let offset = c._offset
+      if offset < 0 { return false }
+
+      switch c {
+      case .remove(_, _, _):
+        if removeOffset.contains(offset) { return false }
+        removeOffset.insert(offset)
+      case .insert(_, _, _):
+        if insertOffset.contains(offset) { return false }
+        insertOffset.insert(offset)
+      }
+
+      if let assoc = c._associatedOffset {
+        if assoc < 0 { return false }
+        switch c {
+        case .remove(_, _, _):
+          if removeOffsetToAssoc[offset] != nil { return false }
+          removeOffsetToAssoc[offset] = assoc
+        case .insert(_, _, _):
+          if insertAssocToOffset[assoc] != nil { return false }
+          insertAssocToOffset[assoc] = offset
+        }
+      }
+    }
+
+    return removeOffsetToAssoc == insertAssocToOffset
+  }
+
+  /// Creates an instance from a collection of changes.
+  ///
+  /// For clients interested in the difference between two collections, see
+  /// `BidirectionalCollection.difference(from:)`.
+  ///
+  /// To guarantee that instances are unambiguous and safe for compatible base
+  /// states, this initializer will fail unless its parameter meets to the
+  /// following requirements:
+  ///
+  /// 1. All insertion offsets are unique
+  /// 2. All removal offsets are unique
+  /// 3. All associations between insertions and removals are symmetric
+  ///
+  /// - Parameter c: A collection of changes that represent a transition
+  ///   between two states.
+  ///
+  /// - Complexity: O(*n* * log(*n*)), where *n* is the length of the
+  ///   parameter.
+  public init?<Changes: Collection>(
+    _ changes: Changes
+  ) where Changes.Element == Change {
+    if !CollectionDifference<ChangeElement>._validateChanges(changes) {
```

Comment:
FWIW: there's an idiomatic way to spell `if !` called `guard`. Some people prefer it for preconditions.

Reply:
Oops, this actually used to be a `guard` but the code got refactored a number of times >_<